[safe templates] Various template fixes.

peter-fogg opened the pull request.

edx-webhook commented:
## PR Cover Letter
- [ ] A detailed description of the changes in the body of the PR
- [ ] If available, a link to the relevant JIRA ticket in the body of the PR
  - `[PROJ-XXXX](https://openedx.atlassian.net/browse/PROJ-XXXX)`

## Sandbox
- [ ] Build a sandbox for your branch and add a link here

## Testing
- [ ] i18n
- [ ] RTL
- [ ] Unit, integration, acceptance tests as appropriate
- [ ] Analytics
- [ ] Performance
- [ ] Database migrations are backwards-compatible

## Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review #1
- [ ] Code review #2
- [ ] Docs review (required for UI strings/error messages)
- [ ] UX review
- [ ] Accessibility review
- [ ] Product review

## Documentation considerations
- [ ] Are there user visible changes? If so, list who is affected, and include a pointer to the relevant code.
- [ ] Should this be mentioned in the release notes? If so, make sure that you included a description.

## DevOps assistance
If it's needed, or even if you suspect it might be needed, do any/all these things as early as possible. Don't surprise devops right before merging! For example, touching envs/aws.py or or envs/common.py are dead giveaways that you'll need assistance when pushing your changes out. Things that require devops attention include, but are not limited to:
- configuration changes
- feature flags
- migrations
Ways to get devops attention, pick-and-choose whichever is most appropriate for your specific PR:
- [ ] Tag devops on your PR (`@edx/devops`), either for review (more specific to devops = better) or just a general FYI
- [ ] File a support ticket

FYI: Tag anyone who might be interested in this PR here.

## Post-review
- [ ] Squash commits

peter-fogg commented:
jenkins run lettuce

robrap commented:
This is a general message being posted for all Safe Templates PRs.  If you are making steady progress, sorry about this, and thank you!

We need to get any Safe Template PRs merged as soon as possible.

Make sure you have the latest copy of `./scripts/safe_template_linter.py` and `./scripts/safe-commit-linter.sh`, and make sure that `./scripts/safe-commit-linter.sh` reports 0 violations for this PR.

Feel free to ping me with any questions, and/or refer to the [Safe Template Linter documentation](http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/conventions/safe_templates.html#safe-template-linter ).

Thanks!

peter-fogg commented:
@AlasdairSwan, would you mind reviewing the Financial Assistance form commit? @andy-armstrong, could you review the other template changes? The bulk of the work is in the common components; the rest of it is mostly just adding the `expression_filter` tag to existing templates.

peter-fogg commented:
For more context, this is a PR from the safe templates marathon that was lying around for a while. I've just rebased it onto master and fixed up a few things, and it would be great to have it merged before the end of the week.

andy-armstrong commented on common/static/common/templates/components/paging-footer.underscore:
```diff
@@ -1,20 +1,20 @@
     <div class="nav-item page">
         <div class="pagination-form">
-            <label class="page-number-label" for="page-number-input"><%= interpolate(
+            <label class="page-number-label" for="page-number-input"><%- interpolate(
```
This should switch to using ``HtmlUtils.interpolate``.

andy-armstrong commented on lms/static/js/commerce/views/receipt_view.js:
```diff
@@ -67,6 +67,7 @@ var edx = edx || {};
             context.username = this.username;
             context.platformName = this.$el.data('platform-name');
+            context.fulfillment_instructions = edx.HtmlUtils.HTML(context.fulfillment_instructions);
             providerDiv.html(_.template(templateHtml)(context)).removeClass('hidden');
```
Switch to using ``HtmlUtils.template``.

andy-armstrong replied:
You will then need to use ``HtmlUtils.setHtml``.

andy-armstrong commented on lms/static/js/financial-assistance/views/financial_assistance_form_view.js:
```diff
@@ -73,7 +74,8 @@
+                        HtmlUtils: HtmlUtils
                     });
 
                     this.$el.html(_.template(this.tpl)(data));
```
Switch to ``HtmlUtils.setHtml`` and ``HtmlUtils.template``.

andy-armstrong commented on lms/static/js/financial-assistance/views/financial_assistance_form_view.js:
```diff
@@ -119,16 +118,16 @@
                             {
-                                link_start: '<a href="' + this.context.account_settings_url + '">',
-                                link_end: '</a>'
+                                link_start: HtmlUtils.HTML(
```
Use ``HtmlUtils.joinHtml`` instead of concatenating strings (or you can use ``HtmlUtils.interpolate`` if you prefer).

andy-armstrong commented on lms/templates/commerce/provider.underscore:
```diff
@@ -1,27 +1,26 @@
 <div class="provider-wrapper">
   <div class="provider-info">
-    <%= interpolate(
+    <%- interpolate(
```
Use ``HtmlUtils.interpolate``.

andy-armstrong commented on lms/templates/commerce/provider.underscore:
```diff
@@ -1,27 +1,26 @@
   </div>
   <div class="provider-more-info">
-    <%= interpolate(
+    <%- interpolate(
```
Use ``HtmlUtils.interpolate``.

andy-armstrong commented on lms/templates/financial-assistance/financial_assessment_form.underscore:
```diff
@@ -36,7 +36,7 @@
+	<%= HtmlUtils.HTML(fieldsHtml) %>
 
 	<div class="cta-wrapper clearfix">
 		<a href="<%- student_faq_url %>" class="nav-link"><%- interpolate_text(
```
Use ``HtmlUtils.interpolate`` instead of ``interpolate_text``.

AlasdairSwan commented:
Once @andy-armstrong's comments have been addressed 👍

andy-armstrong commented:
@peter-fogg Be sure to address all the issues found in the safe_linter report:

https://build.testeng.edx.org/job/edx-platform-quality-pr/19688/artifact/edx-platform/reports/safecommit/safecommit.report/*view*/

andy-armstrong commented:
@peter-fogg Thanks for making the effort to get this over the line before you leave. It looks great, with just some simple issues to address. Let me know when you have it ready for a second review.

peter-fogg commented:
@andy-armstrong I think I've addressed your feedback as well as all the linter's warnings.

andy-armstrong commented on lms/templates/financial-assistance/financial_assessment_form.underscore:
```diff
@@ -39,9 +39,9 @@
-		<a href="<%- student_faq_url %>" class="nav-link"><%- interpolate_text(
-			gettext('Back to {platform} FAQs'),
-    		{platform: platform_name}
+		<a href="<%- student_faq_url %>" class="nav-link"><%- edx.HtmlUtils.interpolateHtml(
```
Nit: You'd don't need to use the edX namespace as ``HtmlUtils.template`` injects ``HtmlUtils`` in as a context variable. If that doesn't work, it means you are using ``_.template`` to render it.

andy-armstrong commented on common/static/common/js/components/views/paging_footer.js:
```diff
@@ -41,7 +41,8 @@
                             total_pages: this.collection.getTotalPages(),
-                            paginationLabel: this.paginationLabel
+                            paginationLabel: this.paginationLabel,
+                            interpolateHtml: HtmlUtils.interpolateHtml,
```
You don't need to do this, as ``HtmlUtils`` is available as a context variable.

andy-armstrong commented on common/static/common/templates/components/paging-footer.underscore:
```diff
@@ -2,10 +2,9 @@
     <div class="nav-item page">
         <div class="pagination-form">
-            <label class="page-number-label" for="page-number-input"><%- interpolate(
+            <label class="page-number-label" for="page-number-input"><%- interpolateHtml(
```
Use ``HtmlUtils.interpolateHtml`` here.

andy-armstrong commented on lms/templates/commerce/provider.underscore:
```diff
@@ -1,14 +1,14 @@
 <div class="provider-wrapper">
   <div class="provider-info">
-    <%- interpolate(
+    <%- edx.HtmlUtils.interpolateHtml(
```
Just use ``HtmlUtils.interpolateHtml``.

andy-armstrong commented on lms/templates/commerce/provider.underscore:
```diff
@@ -1,14 +1,14 @@
   <div class="provider-info">
-    <%- interpolate(
+    <%- edx.HtmlUtils.interpolateHtml(
           gettext("You still need to visit the %(display_name)s website to complete the credit process."),
```
You need to switch ``%(foo)s`` to ``{foo}`` whenever you switch from interpolate to interpolateHtml.

andy-armstrong commented on lms/templates/commerce/provider.underscore:
```diff
@@ -1,14 +1,14 @@
   </div>
   <div class="provider-more-info">
-    <%- interpolate(
+    <%- edx.HtmlUtils.interpolateHtml(
```
Use ``HtmlUtils.interpolateHtml``.

andy-armstrong commented on lms/templates/commerce/provider.underscore:
```diff
@@ -1,14 +1,14 @@
   <div class="provider-more-info">
-    <%- interpolate(
+    <%- edx.HtmlUtils.interpolateHtml(
           gettext("To finalize course credit, %(display_name)s requires %(platform_name)s learners to submit a credit request."),
```
Convert ``%(foo)s`` to ``{foo}``.

robrap replied:
FYI: @andy-armstrong: I added TNL-4775: Lint for %(...)s in HtmlUtils.interpolateHtml() or StringUtils.interpolate() calls

andy-armstrong commented on lms/static/js/commerce/views/receipt_view.js:
```diff
@@ -39,7 +39,7 @@ var edx = edx || {};
             });
 
-            this.$el.html(_.template(templateHtml)(context));
+            edx.HtmlUtils.setHtml(this.$el, _.template(templateHtml)(context));
```
Use ``HtmlUtils.template``.

andy-armstrong commented on lms/templates/commerce/checkout_cancel.html:
```diff
@@ -8,6 +12,6 @@
-    ${ _(u"Your transaction has been cancelled. If you feel an error has occurred, contact {email}.").format(
-    email="<a href=\"mailto:{email}\">{email}</a>".format(email=payment_support_email)) }
+    ${ Text(_(u"Your transaction has been cancelled. If you feel an error has occurred, contact {email}.")).format(
+    email=HTML("<a href=\"mailto:{email}\">{email}</a>").format(email=payment_support_email)) }
```
Nit: can you clean up the indentation (I realize it wasn't you who wrote it this way). Thanks.

andy-armstrong commented on lms/templates/commerce/checkout_error.html:
```diff
@@ -1,13 +1,20 @@
+  <h1>${_("Checkout Error")}</h1>
+  ## Translators: bold_end and bold_start are HTML tags to create bold text.
+  ${ Text(_(u"An error has occurred with your payment. {bold_start}You have not been charged.{bold_end} Please try to submit your payment again. If this problem persists, contact {email}.")).format(
+  email=HTML("<a href=\"mailto:{email}\">{email}</a>").format(email=payment_support_email),
```
Nit: indent the parameters four spaces, and put the trailing paren on a new line.

andy-armstrong commented on lms/static/js/financial-assistance/views/financial_assistance_form_view.js:
```diff
@@ -87,7 +87,7 @@
 
                 renderSuccess: function() {
-                    this.$el.html(_.template(successTpl)({
+                    HtmlUtils.setHtml(this.$el, _.template(successTpl)({
```
Use ``HtmlUtils.template``.

andy-armstrong commented on lms/static/js/financial-assistance/views/financial_assistance_form_view.js:
```diff
@@ -131,7 +131,7 @@
                     if( !this.model.get('country') ){
                         $countryLabel.addClass('error');
-                        $errorMessageContainer.append("<li>" + msg + "</li>");
+                        HtmlUtils.append($errorMessageContainer, HtmlUtils.joinHtml("<li>", msg, "</li>"));
```
This is now escaping ``msg`` when it wasn't before. It is also going to be escaping the HTML strings. This needs to be more like:

    HtmlUtils.joinHtml(
        HtmlUtils.HTML('<li>'), 
        HtmlUtils.HTML(msg),
        HtmlUtils.HTML('</li>')
    )

andy-armstrong commented:
@peter-fogg This is tricky to get right! Here are some more comments so let me know when you want me to look at it again.

robrap commented on lms/templates/commerce/provider.underscore:
```diff
@@ -1,13 +1,13 @@
   <div class="provider-info">
-    <%- edx.HtmlUtils.interpolateHtml(
-          gettext("You still need to visit the %(display_name)s website to complete the credit process."),
+    <%- HtmlUtils.interpolateHtml(
```
For these two changes where you aren't mixing in HTML tags, just replace `HtmlUtils.interpolateHtml` with `StringUtils.interpolate`, otherwise you'd be double-escaping.

robrap commented on lms/templates/commerce/provider.underscore:
```diff
@@ -1,27 +1,26 @@
-    <%= interpolate(
-          gettext("You still need to visit the %(display_name)s website to complete the credit process."),
-          { display_name: display_name }, true
+    <%- HtmlUtils.interpolateHtml(
```
You can replace `HtmlUtils.interpolateHtml` with `StringUtils.interpolate` here, otherwise you'd be double-escaping.

robrap commented on lms/templates/commerce/provider.underscore:
```diff
@@ -1,27 +1,26 @@
-    <%= interpolate(
-          gettext("To finalize course credit, %(display_name)s requires %(platform_name)s learners to submit a credit request."),
-          { display_name: display_name, platform_name: platformName }, true
+    <%- HtmlUtils.interpolateHtml(
```
Same as above.

robrap commented on lms/templates/commerce/provider.underscore:
```diff
@@ -1,27 +1,26 @@
   </div>
   <div class="provider-instructions">
-    <%= fulfillment_instructions %>
+    <%- edx.HtmlUtils.ensureHtml(fulfillment_instructions) %>
```
Either `fulfillment_instructions` contains HTML, in which case you'd want it to be an HtmlSnippet first, and then use:
```
<%= edx.HtmlUtils.ensureHtml(fulfillment_instructions) %>
```
Or, it does not include HTML, and you would want:
```
<%- fulfillment_instructions %>
```

andy-armstrong replied:
This can be just ``HtmlUtils``, not ``edx.HtmlUtils``.

robrap commented on lms/templates/financial-assistance/financial_assessment_form.underscore:
```diff
@@ -36,12 +36,12 @@
-		<a href="<%- student_faq_url %>" class="nav-link"><%- interpolate_text(
-			gettext('Back to {platform} FAQs'),
-    		{platform: platform_name}
+		<a href="<%- student_faq_url %>" class="nav-link"><%- HtmlUtils.interpolateHtml(
```
Again, use `StringUtils` and let the `<%-` take care of the escaping.

andy-armstrong commented on lms/templates/financial-assistance/financial_assessment_form.underscore:
```diff
@@ -39,7 +39,7 @@
 
 	<div class="cta-wrapper clearfix">
-		<a href="<%- student_faq_url %>" class="nav-link"><%- edx.HtmlUtils.interpolateHtml(
+		<a href="<%- student_faq_url %>" class="nav-link"><%- HtmlUtils.interpolateHtml(
```
I think this should use ``StringUtils.interpolate`` since it isn't producing HTML. It is a red flag for me to use ``<%-`` on the result of ``HtmlUtils.interpolateHtml`` because that will end up double-escaping.

robrap commented on lms/templates/support/certificates.html:
```diff
@@ -9,8 +12,8 @@
-        userFilter: '${ user_filter }',
-        courseFilter: '${course_filter}'
+        userFilter: '${ user_filter | n, js_escaped_string}',
+        courseFilter: '${ course_filter | n, js_escaped_string }'
```
Nit: Drop surrounding spaces.  At this time it is inconsistent in userFilter as well.
Use:
```
 userFilter: '${user_filter | n, js_escaped_string}',
 courseFilter: '${course_filter | n, js_escaped_string}'
```

andy-armstrong commented:
@peter-fogg This looks really close. Do you have a sandbox to verify these changes, as it is very easy to accidentally double escape something (we had a lot of rollbacks after the safe template marathon for this reason).

robrap commented on common/static/common/templates/components/paging-footer.underscore:
```diff
@@ -1,20 +1,19 @@
     <div class="nav-item page">
         <div class="pagination-form">
-            <label class="page-number-label" for="page-number-input"><%= interpolate(
+            <label class="page-number-label" for="page-number-input"><%- HtmlUtils.interpolateHtml(
```
Use `StringUtils.interpolate` instead, and switch next line to:
```
gettext("Page number out of {total_pages}"),
```

peter-fogg commented:
@andy-armstrong Can you take another (hopefully last?) look?

andy-armstrong commented on lms/static/js/financial-assistance/views/financial_assistance_form_view.js:
```diff
@@ -10,7 +10,7 @@
             'text!../../../templates/financial-assistance/financial_assessment_submitted.underscore',
             'text!templates/student_account/form_field.underscore',
-            'string_utils'
+            'edx-ui-toolkit/js/utils/html-utils'
```
Nit: I prefer putting ``HtmlUtils`` earlier in the list, probably after ``gettext``. I feel that the dependencies should be the generic helpers first, and then the more specific libraries later.

peter-fogg replied:
Done.

andy-armstrong commented on lms/templates/financial-assistance/financial_assessment_form.underscore:
```diff
@@ -36,12 +36,12 @@
 	</div>
 
-	<%= fields %>
+	<%= HtmlUtils.HTML(fieldsHtml) %>
```
This should be ``HtmlUtils.ensureHtml``, if I remember correctly. I think this will double escape as it is written.

peter-fogg replied:
Done.

robrap replied:
@andy-armstrong: `ensureHtml()` is safer and the better choice because it ensures the fieldsHtml is properly escaped, whether it is an HtmlSnippet or not.  However, `HTML()` shouldn't double-escape, because it is a way for you to simply declare something as containing HTML, which would hopefully be used only when you know you have safe HTML that was HTML before HtmlUtils could have gotten involved.

andy-armstrong commented on common/static/common/js/components/views/paging_footer.js:
```diff
@@ -41,7 +41,7 @@
                             current_page: this.collection.getPageNumber(),
                             total_pages: this.collection.getTotalPages(),
-                            paginationLabel: this.paginationLabel
+                            paginationLabel: this.paginationLabel,
```
Nit: the extra comma isn't valid in JS.

andy-armstrong commented on common/static/common/templates/components/paging-footer.underscore:
```diff
@@ -1,20 +1,19 @@
         <div class="pagination-form">
-            <label class="page-number-label" for="page-number-input"><%= interpolate(
+            <label class="page-number-label" for="page-number-input"><%- StringUtils.interpolate(
                     gettext("Page number out of %(total_pages)s"),
```
This needs to be switched to ``{total_pages}``.

andy-armstrong commented on common/static/common/templates/components/paging-footer.underscore:
```diff
@@ -1,20 +1,19 @@
 
-        <span class="current-page"><%= current_page %></span>
+        <span class="current-page"><%- current_page %></span>
         <span class="sr">&nbsp;out of&nbsp;</span>
```
Wow, is this not using ``gettext``?

andy-armstrong replied:
I think this is probably out of scope for your PR. This whole section needs to use sentences for the translators, as it isn't possible to translate snippets like this.

andy-armstrong commented on lms/static/js/commerce/views/receipt_view.js:
```diff
@@ -39,7 +39,7 @@ var edx = edx || {};
             });
 
-            this.$el.html(_.template(templateHtml)(context));
+            edx.HtmlUtils.setHtml(this.$el, edx.HtmlUtils.template(templateHtml)(context));
```
Nit: if you have time, it might be cleaner to pass ``edx.HtmlUtils`` to the anonymous outer function, so you don't have to duplicate the namespace everywhere. This is very minor so only do it if you're changing the code anyway.

andy-armstrong commented on lms/templates/commerce/provider.underscore:
```diff
@@ -1,27 +1,26 @@
 <div class="provider-buttons-logos">
   <div class="provider-logo">
-    <%= interpolate("<img src='%s' alt='%s'></image>", [thumbnail_url, display_name]) %>
+    <img src="<%- thumbnail_url %>" alt="<%- display_name %>" />
```
This is so much easier to read without ``interpolate``! It shows one of the benefits of doing this logic in Underscore and not in JavaScript.

andy-armstrong commented:
@peter-fogg This is just down to nits now. Thanks so much for following through on this! 👍

peter-fogg commented:
jenkins run python

peter-fogg commented:
jenkins run bokchoy

peter-fogg commented:
jenkins run bokchoy

edx-webhook commented:
Hello @peter-fogg: We are unable to continue with review of your submission at this time. Please see the associated JIRA ticket for more explanation.
